### DISCUSSION TITLE
0.13.0-stable has been released!

### DISCUSSION THREAD

**Discussion by mihe** [godot-jolt/godot-jolt#939] [Announcements]
This release only supports **Godot 4.3** (starting with [4.3-stable](https://godotengine.org/releases/4.3/)).

Note that this release, unlike the one found in the Godot Asset Library, includes double-precision builds, titled `*_double`. These are only compatible with double-precision builds of Godot (which must be built from source) and won't load at all in the regular (single-precision) releases of Godot. They've also not been very thoroughly tested, so [please report](https://github.com/godot-jolt/godot-jolt/issues/new) any issue you run into when using them. 

## Changelog

These are the notable changes that have been made since [0.12.0-stable](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.12.0-stable) was released. You can also find a list of all the commits [here](https://github.com/godot-jolt/godot-jolt/compare/v0.12.0-stable...v0.13.0-stable).

Any breaking changes are denoted with ⚠️.

### Removed

- ⚠️ Removed any `JoltPhysicsServer3D.G6DOF_*` enums that were previously (i.e. before Godot 4.3) not exposed by Godot itself. These can now be found as `PhysicsServer3D.G6DOF_*` instead.

### Changed

- ⚠️ Changed the way that the `body_test_motion` method of `PhysicsServer3D` discards contacts, which should fix issues related to jitter/ping-ponging. This affects `move_and_slide`, `move_and_collide` and `test_move`.
- ⚠️ Changed the inertia of shapeless bodies to be `(1, 1, 1)`, to match Godot Physics.
- Changed `SeparationRayShape3D` to not treat other convex shapes as solid, meaning it will now only ever collide with the hull of other convex shapes, which better matches Godot Physics.
- Changed so that modifying joint properties now wake up any bodies connected to them.

### Added

- Added support for `SoftBody3D`.
- Added support for double-precision.
- Added `compatibility_maximum` to `godot-jolt.gdextension` for controlling the maximum allowed Godot version. Note that changing this may expose you to subtle (or not so subtle) incompatibilities in the extension API. Change at your own risk.
- ⚠️ Added new project setting, "Use Enhanced Internal Edge Detection", which can help alleviate collisions with internal edges of `ConcavePolygonShape3D` and `HeightMapShape3D` shapes, also known as ghost collisions. This setting is enabled by default and may change the behavior of character controllers relying things like `move_and_slide` or `move_and_collide`.
- Added support for partial custom inertia, where leaving one or two components at zero will use the automatically calculated values for those specific components.
- Added error-handling for invalid scaling of bodies/shapes.
- Added project settings "Body Pair Cache Enabled", "Body Pair Cache Distance Threshold" and "Body Pair Cache Angle Threshold" to allow fine-tuning the scale by which collision results are reused inbetween physics ticks.
- Added the appropriate floating-point precision feature tag to `godot-jolt.gdextension` to prevent accidentally loading a single-precision build of Godot Jolt in a double-precision build of Godot and vice versa.

### Fixed

- ⚠️ Fixed issue with shape queries not returning the full contact manifold. This applies to the `collide_shape` method of `PhysicsDirectSpaceState3D` as well as the `body_test_motion` method of `PhysicsServer3D`, which subsequently affects the `test_move` and `move_and_collide` methods of `PhysicsBody3D` as well as the `move_and_slide` method of `CharacterBody3D`.
- ⚠️ Fixed issue with the `body_get_direct_state` method of `PhysicsServer3D` returning a non-null reference when the body has no space.
- Fixed issue with not being able to pass a physics space `RID` to `area_get_param`, `area_attach_object_instance_id` and `area_get_object_instance_id`.
- Fixed issue where the `inverse_inertia` property of `PhysicsDirectBodyState3D` would have some of its components swapped.
- Fixed issue where shapeless bodies wouldn't have custom center-of-mass applied to them.
- Fixed issue with high (>1) damping values producing different results across different update frequencies.
- Fixed issue where physics queries performed within the editor (e.g. editor plugins or tool scripts) would end up ignoring potentially large swathes of bodies in scenes with many physics bodies.
- Fixed crash when setting a negative `max_contacts_reported` on `RigidBody3D`.
- Fixed issue with certain `generic_6dof_joint_*` methods on `JoltPhysicsServer3D` not having the correct parameter names.
- Fixed issue where sleep timer would not reset when applying force/torque to a `RigidBody3D`.
- Fixed issue where `Area3D` would fail to report overlap with certain `ConcavePolygonShape3D`.
- Fixed crash when re-enabling the node processing of a disabled body that's connected to a two-body joint.
- Fixed issue where re-adding previously used physics bodies to the scene tree could result in many "Unhandled override mode" errors in the Output log.
- Fixed memory leak when changing the data/properties of `CapsuleShape3D`, `CylinderShape3D`, `SeparationRayShape3D`, `ConcavePolygonShape3D`, and `HeightMapShape3D`.

<hr /><em>This discussion was created from the release <a href='https://github.com/godot-jolt/godot-jolt/releases/tag/v0.13.0-stable'>0.13.0-stable</a>.</em>

**Comment by kookooyama** [comment]
Hello, thanks for the update.

I have a problem after updating to 4.3 and getting this version of Jolt. I am now getting flooded with errors when playing. These errors don't seem to break anything but is there a way to suppress them? Here is the error that I am getting:

try_build_shape: Failed to correctly scale body 'TestArea:<Area3D#93935634318>'. A scale of (-0.875956, -0.805722, -0.997483) is not supported by Godot Jolt for this shape/body. The scale will instead be treated as (-0.893054, -0.893054, -0.893054). <C++ Source>   src\objects\jolt_shaped_object_impl_3d.cpp:140 @ try_build_shape()

Thank you.

**Reply by mihe** [reply]
> These errors don't seem to break anything

You will end up with collision detection for that `Area3D` that is likely quite a ways off from what you actually want, so I would say it does break something. The fact that your scale is negative might also be a problem, if that wasn't intentional.

> but is there a way to suppress them?

There isn't a way to suppress them per se, but you can call `orthonormalize()` at the end of `_physics_process` of your `Area3D` if you really don't care. That will normalize the axes of your transforms, and thus your scale, which should be valid regardless of shape type/arrangement.

**Reply by kookooyama** [reply]
I scale the X at 1 or -1 to mirror the character, so that's why the scale can end up being negative. I also have the same issue with any PhysicalBone3D. Thanks for the suggestion, I'll try to use orthonormalize if I absolutely need to but based on what you've said I think some refactoring seems to be in order to not break anything else in the future.

**Comment by Gamepro5** [comment]
Crazy speedrun for support of 4.3 :)

**Comment by fatduckling** [comment]
Thanks very much @mihe
